ocaml/xcp-rrdd/scripts/rrdd/rrdd.py

```diff
@@ -310,20 +311,27 @@ def update(self):
         metadata_json = json.dumps(metadata, sort_keys=True).encode('utf-8')
         metadata_checksum = crc32(metadata_json) & 0xffffffff
 
-        self.dest.seek(0)
-        self.dest.write('DATASOURCES'.encode())
-        self.dest.write(pack(">LLLQ",
-                             data_checksum,
-                             metadata_checksum,
-                             len(self.datasources),
-                             timestamp))
+        # First write the updated data and metadata
+        encoded_datasource_header = 'DATASOURCES'.encode()
+        # DATASOURCES + 20 for 32 + 32 + 32 + 64
+        self.dest.seek(len(encoded_datasource_header) + 20)
         for val in data_values:
             # This is already big endian encoded
             self.dest.write(val)
```

Comment:
The reader can do its reading at any time? Is it reasonable to assume that this action of writing the the data/metadata is atomic?

Reply:
the reader should not read any more than the header unless and until the `timestamp` in the header changes from the last data read. Once that changes the data can be read.

Reply:
Header and data is protected by checksums. So if you read garbage, the checksum should tell you. You can't avoid reading garbage in the general case because there is always a race between the data and its checksum being written as no locking is implemented.

The total amount of data can change, and in particular shrink. I believe the memory-mapped file is of constant size though.

Reply:
Ok, I think at least it is better to write the data before writing the header, that sounds like a safer appraoch than the other way round